### ISSUE TITLE
Can StructScan decode to an interface given the reflect.Type and a table name?

### ISSUE THREAD

**Issue by mbucc** [jmoiron/sqlx#125]
Hi,

I'm trying to use sqlx like gob; I want register concrete types and then want "decode" them from SQL back into a slice of interfaces.  For example, I want to transform this:

```go

type Event interface{
	ID() int
}

type S struct {
	Id int
}

func (s *S) ID() int {
	return s.Id
}

func add(s reflect.Value, idx int) {
	
	entry := reflect.New(reflect.TypeOf(S{}))
	field := entry.Elem().FieldByIndex([]int{0})
	field.SetInt(int64(idx))
	
	s.Index(idx).Set(entry)
}

func main() {
	var a []Event

	s := reflect.ValueOf(&a).Elem()
	s.Set(reflect.MakeSlice(reflect.TypeOf(a), 2, 2))
	
	add(s, 0)
	add(s, 1)

	fmt.Println(a[0], a[1])
}
```

into somthing that uses sqlx in the ```add()``` function.  This is the closest I could get:

```go
func add(s reflect.Value, idx int) {
	rows, _ := db.Queryx("select * from [main.S]")
	for rows.Next() {
		v0 := reflect.New(reflect.TypeOf(S{}))
		v1 := s0.Elem().Interface()
		_ = rows.StructScan(&s)
		s.Index(idx).Set(v0)
		idx++
	}
}
```

(This panics with "TraversalsByName on interface Value")

Is this possible?

Thanks!

(Working example of first snippet: https://play.golang.org/p/EKZSxKvt3i).

**Comment by jmoiron** [comment]
I don't quite know what you're trying to do.  Your original code seems to be basically re-implementing the `append` builtin.

Your sqlx sample code has many, many errors in it;  why pass a reflect.Value if you're just going to take the `TypeOf(S{})`?  Why are you trying to scan into a `reflect.Value`?  Where is the `s0` variable coming from?

That said, you should be able to scan into an `interface{}` if it has an addressable struct inside it (without passing a pointer to it;  pointer to an interface isn't what you want).

If you want to select a bunch of events from SQL, you could easily just do

```go
var a []Event
err := db.Select(&a, "SELECT * FROM events;")
```

**Comment by mbucc** [comment]
I'm so deep in the weeds on this one, I'm not surprised my intent wasn't clear.  Let me rephrase.

Given:
* a sql statement that selects all MyEvent from a table
* an interface Event
* the reflect.Type of MyEvent (which implements the interface Event)

I am trying to write this function:
```go
func getEvents(sqlx.DB *db, sql string, typ reflect.Type) []Event {
}
```

The original code I posted is a boiled down version of what encoding/gob does when it decodes serialized structs into an slice of interfaces, which is pretty much the same problem I am trying to tackle with sqlx.

It would be awesome if I am missing something really simple ... :)

**Comment by mbucc** [comment]
This does what I want.

```go
func test12(db *sqlx.DB) {
	ptr := reflect.New(reflect.SliceOf(reflect.TypeOf(&MyEvent{})))
	iface := ptr.Interface()
	err := db.Select(iface, "select * from [main.MyEvent]")
	if err != nil {
		panic(err)
	}
	x0 := ptr.Elem().Index(0)
	var i Event = x0.Interface().(Event)
}```

* Event is the interface
* &MyEvent implements the interface
* all we have to work with is TypeOf(&MyEvent) and Event.

Closing ticket, thanks.
